### PR TITLE
[Draft] Enable rocMLIR on Windows

### PR DESCRIPTION
**Pull request by apwojcik**
That draft PR illustrates the changes required to build and run rocMLIR on Windows. The changes were tested on Linux, however, they are not yet in the CI.

More details on the effort will come soon.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -66,9 +72,14 @@ endif()
 # Set up the build for the LLVM/MLIR git-submodule
 include(cmake/llvm-project.cmake)
 
-# Update the build-tree RPATH
-set(CMAKE_BUILD_RPATH "${ROCMLIR_LIB_DIR};${LLVM_EXTERNAL_LIB_DIR}")
-message(STATUS "CMAKE_BUILD_RPATH: ${CMAKE_BUILD_RPATH}")
+if(MSVC)
```
Comment (krzysz00) [inline CMakeLists.txt:75]: Hold on just a moment - why isn't this stuff being pulled in from upstream's build setup? LLVM builds on Windows, after all?

Also, if this is all duplicative of upstream, can it be moved to `mlir/CMakeLists.txt`?

File: external/CMakeLists.txt [hunk 2]
```diff
@@ -1 +1,2 @@
+add_definitions(-D_SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING)
```
Comment (krzysz00) [inline external/CMakeLists.txt:1]: Why do you need this? Should it perhaps, instead, be a patch to LLVM? If it is already fixed upstream, can you pull in that patch as a backport?

File: external/llvm-project/mlir/test/CMakeLists.txt [hunk 3]
```diff
@@ -10,6 +10,14 @@ endif()
 set(MLIR_CMAKE_DIR
   "${CMAKE_BINARY_DIR}/lib${LLVM_LIBDIR_SUFFIX}/cmake/mlir")
 
+if(WIN32)
```
Comment (krzysz00) [inline external/llvm-project/mlir/test/CMakeLists.txt:13]: These are patches to upstream - are you willing to take responsibility for upstreaming them?

Furthermore, changes to `external/llvm-project/` should be separate from other changes for git subtree reasons

File: external/llvm-project/mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-addi-i16.mlir [hunk 4]
```diff
@@ -4,14 +4,14 @@
 // RUN: mlir-opt %s --convert-scf-to-cf --convert-cf-to-llvm --convert-vector-to-llvm \
 // RUN:             --convert-func-to-llvm --convert-arith-to-llvm | \
 // RUN:   mlir-cpu-runner -e entry -entry-point-result=void \
-// RUN:                   --shared-libs=%mlir_lib_dir/libmlir_c_runner_utils%shlibext | \
+// RUN:                   --shared-libs=%linalg_test_lib_dir/%prefix_mlir_c_runner_utils%shlibext | \
```
Comment (krzysz00) [inline external/llvm-project/mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-addi-i16.mlir:7]: Ok, having looked, what you actually want is to pull https://reviews.llvm.org/D143925 onto our repository instead of reinventing it badly

Comment (krzysz00) [reply external/llvm-project/mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-addi-i16.mlir:7]: (That might not be the exact commit but still)

File: mlir/CMakeLists.txt [hunk 5]
```diff
@@ -8,8 +8,6 @@ set(LLVM_LIBRARY_OUTPUT_INTDIR ${CMAKE_BINARY_DIR}/lib)
 set(MLIR_BINARY_DIR ${CMAKE_BINARY_DIR})
 set(MLIR_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
 
-list(APPEND CMAKE_MODULE_PATH "${MLIR_CMAKE_DIR}")
```
Comment (krzysz00) [inline mlir/CMakeLists.txt:11]: Why did you remove these?

File: mlir/lib/ExecutionEngine/CMakeLists.txt [hunk 6]
```diff
@@ -6,8 +6,10 @@ target_link_libraries(conv-validation-wrappers
...
-# that apparently works out okay.
-set_source_files_properties(conv-validation-wrappers.cpp PROPERTIES
-  COMPILE_OPTIONS "-Wno-gnu-anonymous-struct;-Wno-nested-anon-types;-Wno-return-type-c-linkage;-Wno-c++98-compat-extra-semi")
+if (NOT MSVC)
+  # Suppress some compiler warnings when building with clang.  The first two
+  # suppressions regard a GNU extension, the third an incomplete-type warning
+  # that apparently works out okay.
+  set_source_files_properties(conv-validation-wrappers.cpp PROPERTIES
+    COMPILE_OPTIONS "-Wno-gnu-anonymous-struct;-Wno-nested-anon-types;-Wno-return-type-c-linkage;-Wno-c++98-compat-extra-semi")
+endif()
```
Comment (krzysz00) [inline mlir/lib/ExecutionEngine/CMakeLists.txt:15]: Nit: newline

File: mlir/utils/performance/ck-benchmark-driver/CMakeLists.txt [hunk 7]
```diff
@@ -1,17 +1,9 @@
-find_package(composable_kernel CONFIG 1.0.0 COMPONENTS device_operations)
-if(NOT composable_kernel_FOUND)
-  message(WARNING "Could not find composable-kernels, not building a CK benchmark driver")
-endif()
-
+find_package(composable_kernel 1.0.0 COMPONENTS device_operations CONFIG)
 if (composable_kernel_FOUND)
-  find_package(hip REQUIRED PATHS /opt/rocm)
-  set(LIBS composable_kernel::device_operations hip::device benchmark-driver-utils)
-  set(EXCLUDE_FROM_ALL TRUE)
```
Comment (krzysz00) [inline mlir/utils/performance/ck-benchmark-driver/CMakeLists.txt:9]: Please explain these changes 
Also, what's wrong with PATHS - if it's not in the path, it'll look somewhere else, no?

File: mlir/utils/performance/CMakeLists.txt [hunk 8]
```diff
@@ -15,7 +15,7 @@ set(PERFORMANCE_SCRIPTS
   tuningRunner.py
 )
 
-list(TRANSFORM PERFORMANCE_SCRIPTS PREPEND "${CMAKE_CURRENT_SOURCE_DIR}/")
```
Comment (krzysz00) [inline mlir/utils/performance/CMakeLists.txt:18]: Why remove the quotes? Won't they be needed in case the directory contains a space?

File: mlir/tools/rocmlir-lib/CMakeLists.txt [hunk 9]
```diff
@@ -34,7 +36,7 @@ mlir_check_link_libraries(rocmlir-lib-test)
 llvm_canonicalize_cmake_booleans(BUILD_FAT_LIBROCKCOMPILER)
 # Static library target, enabled only when building static libs
 if(BUILD_FAT_LIBROCKCOMPILER)
-  find_package(ROCM 0.8 REQUIRED PATHS /opt/rocm)
```
Comment (krzysz00) [inline mlir/tools/rocmlir-lib/CMakeLists.txt:37]: Why this removal?

File: mlir/tools/rocmlir-lib/CMakeLists.txt [hunk 10]
```diff
@@ -45,57 +47,344 @@ if(BUILD_FAT_LIBROCKCOMPILER)
   set(PACKAGE_NAME ${CMAKE_PROJECT_NAME_LOWER})
 
   set(LIBRARY_NAME rockCompiler)
-  function(combine_archives output_archive)
-    set(mri_file ${CMAKE_CURRENT_BINARY_DIR}/${output_archive}.mri)
-    set(full_output_path ${ROCMLIR_LIB_DIR}/lib${output_archive}.a)
-    set(output_archive_dummy_file ./${output_archive}.dummy.cpp)
+  set(COMPONENT_NAME ${LIBRARY_NAME})
+
+  set(__rocmlir_targets
```
Comment (krzysz00) [inline mlir/tools/rocmlir-lib/CMakeLists.txt:52]: Reject for please find a way to not hard-code this. This'll be failure-prone and it's unclear why we needed it.

**Review by krzysz00 — CHANGES_REQUESTED** [review]
Re the changes to upstream: please consider adopting the approach that's been taken over the last few months there instead of giving us a bunch of merge conflicts in a few weeks. 
If you're AMD-internal, https://confluence.amd.com/pages/viewpage.action?pageId=657605094#Submittingpatchesupstream(andperformingbackports)-Backports has the magic incantations for cherry-picking from upstream MLIR.

The changes to packaging feel like a massive hack, let's find a way to not do them.

And whatever upstream's done to make integration tests work on Windows should be usable for our internal tests as well.

File: mlir/tools/rocmlir-lib/CMakeLists.txt (hunk 10 above)
Comment (apwojcik) [reply mlir/tools/rocmlir-lib/CMakeLists.txt:52]: While working on the Windows port, we discovered `file(GLOB..)` pulls more than required, polluting the fat library with unwanted `*.lib` files.

Comment (krzysz00) [reply mlir/tools/rocmlir-lib/CMakeLists.txt:52]: Also, as a note, we've started a sync up with upstream, so the changes you had to make to `external/` shouldn't be needed anymore ... and will give you a pile of merge conflicts.

Second, please explain more about how those files were unwanted? And can you instead *remove* them from the list (perhaps with some more regex matching on the list element) instead of doing this, which, again, looks like something that'll get tedious to maintain and/or blow up in our face at the worst possible moment.

**Comment by apwojcik** [comment]
The changes to packaging are essential on Windows. Otherwise, we cannot build Debug version of rocMLIR. The size limit of a `*.lib` file is **4GB**. The Debug version of the fat library exceeds that limit. Hence, we must hack it with the INTERFACE library to make it transparent for MIOpen and MIGraphX. It is only affecting Windows, Linux has not been changed.

**Comment by apwojcik** [comment]
Closing due to opening #1133
